gc/base/ParallelDispatcher.cpp

```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+			_statusTable[index] = worker_status_inactive;
+			_threadTable[index] = NULL;
+		}
+
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = _threadCount = _threadCountMaximum = newThreadCount;
+		_activeThreadCount = adjustThreadCount(_threadCount);
+
+		_inShutdown = false;
+
+		TRIGGER_J9HOOK_MM_PRIVATE_GC_THREAD_POOL_ADJUSTED(_extensions->privateHookInterface, env->getOmrVMThread(), J9HOOK_MM_PRIVATE_GC_THREAD_POOL_ADJUSTED, _extensions->gcThreadCount);
+		Trc_MM_ParallelDispatcher_criu_contractThreadPool_Success(preShutdownThreadCount, newThreadCount);
+	}
+
+	Trc_MM_ParallelDispatcher_criu_contractThreadPool_Exit(_extensions->gcThreadCount);
+}
+
+/**
+ * Expand/fill the thread pool by starting up newThreadCount amount of threads. This API is capped by the initial
+ * thread pool size, i.e we can't expand past _threadCountAtGCInit (currently don't support expanding dispatcher tables).
+ *
+ * This API assumes  CRIU is the only consumer (not tested for general use).
+ * The following conditions are required while expanding the thread pool:
+ *     1) NOT holding exclusive VM access
+ *     2) Dispatcher is idle (no task can be dispatched)
+ *     3) Can't be in/enter shutdown
+ */
+bool
+MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env, uintptr_t newThreadCount)
+{
+	Trc_MM_ParallelDispatcher_criu_expandThreadPool_Entry();
+
+	Assert_MM_false(_workerThreadsReservedForGC);
+	Assert_MM_false(_inShutdown);
+
+	Assert_MM_true(_threadShutdownCount == (_threadCountMaximum - 1));
+	Assert_MM_true(_threadCountMaximum == _extensions->gcThreadCount);
+
+	bool result = true;
+
+	uintptr_t supportedGCThreadCount = _extensions->configuration->supportedGCThreadCount(env);
+	uintptr_t preRestoreThreadCount = _extensions->gcThreadCount;
+
+	/* Attempt to use H/W supported Thread Count if
+	 * 1) User didn't specify a restore thread count (restoreGCthreadCount == UDATA_MAX), or
+	 * 2) User specified a restore thread count, but it's greater than our hard limit (_threadCountAtGCInit) */
+	if (newThreadCount > _threadCountAtGCInit) {
+		newThreadCount = supportedGCThreadCount;
+	}
+
+	newThreadCount = OMR_MIN(_threadCountAtGCInit, newThreadCount);
+
+	Trc_MM_ParallelDispatcher_criu_expandThreadPool_params(_extensions->restoreGCthreadCount, newThreadCount, _threadCountAtGCInit, supportedGCThreadCount, preRestoreThreadCount);
+
+	if (newThreadCount > preRestoreThreadCount) {
+
+		result = internalStartupThreads(preRestoreThreadCount, newThreadCount);
+
+		if (result) {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Success();
+			Assert_MM_true(_threadShutdownCount == (newThreadCount - 1));
+		} else {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed(_threadShutdownCount);
+			/* We don't know how many threads started up prior to failing, infer from _threadShutdownCount */
+			newThreadCount = _threadShutdownCount + 1;
```

Comment:
Should we return false and force failure when we fail to startup all threads, or just continue on (as this code does)?

Currently, during VM startup, if this occurs we shutdown all threads and return false to caller which handles it by returning error.

```
	bool result = internalStartupThreads(workerThreadCount, _threadCountMaximum);

	if (result) {
		_threadCount = _threadCountMaximum;
		_activeThreadCount = adjustThreadCount(_threadCount);
	} else {
		/* Clean up the thread table and monitors */
		shutDownThreads();
	}
```

Reply:
In OpenJ9, it seems to shutdown collector and report an error 

```C++
int
gcStartupHeapManagement(J9JavaVM *javaVM)
...

	if (!extensions->dispatcher->startUpThreads()) {
		extensions->dispatcher->shutDownThreads();
		result = JNI_ENOMEM;
	}

	if (JNI_OK != result) {
		PORT_ACCESS_FROM_JAVAVM(javaVM);
		extensions->getGlobalCollector()->collectorShutdown(extensions);
		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_FAILED_TO_STARTUP_GARBAGE_COLLECTOR);
		return result;
	}
```

Reply:
do the same, shutdown threads and return error to the caller (which probably should continue with failure path, too)

Reply:
Will leave this code as is, i.e not assume we'll fail and leave it to the caller to take action based on the `false` value returned.

CRIU code will be changed: https://github.com/eclipse-openj9/openj9/pull/16385#issuecomment-1333142859